src/nxt_router_access_log.c

```diff
@@ -190,25 +274,105 @@ nxt_router_access_log_writer(nxt_task_t *task, nxt_http_request_t *r,
 
     ctx->access_log = access_log;
 
-    if (nxt_tstr_is_const(format->tstr)) {
-        nxt_tstr_str(format->tstr, &ctx->text);
+    if (format->tstr != NULL) {
+        ret = nxt_router_access_log_text(task, r, ctx, format->tstr);
+
+    } else {
+        ret = nxt_router_access_log_json(task, r, ctx, format);
+    }
+
+    if (ret == NXT_OK) {
+        nxt_router_access_log_write(task, r, ctx);
+    }
+}
+
+
+static nxt_int_t
+nxt_router_access_log_text(nxt_task_t *task, nxt_http_request_t *r,
+    nxt_router_access_log_ctx_t *ctx, nxt_tstr_t *tstr)
+{
+    nxt_int_t          ret;
+    nxt_router_conf_t  *rtcf;
+
+    if (nxt_tstr_is_const(tstr)) {
+        nxt_tstr_str(tstr, &ctx->text);
 
     } else {
         rtcf = r->conf->socket_conf->router_conf;
 
         ret = nxt_tstr_query_init(&r->tstr_query, rtcf->tstr_state,
                                   &r->tstr_cache, r, r->mem_pool);
         if (nxt_slow_path(ret != NXT_OK)) {
-            return;
+            return NXT_ERROR;
         }
 
-        ret = nxt_tstr_query(task, r->tstr_query, format->tstr, &ctx->text);
+        ret = nxt_tstr_query(task, r->tstr_query, tstr, &ctx->text);
         if (nxt_slow_path(ret != NXT_OK)) {
-            return;
+            return NXT_ERROR;
         }
     }
 
-    nxt_router_access_log_write(task, r, ctx);
+    return NXT_OK;
+}
+
+
+static nxt_int_t
+nxt_router_access_log_json(nxt_task_t *task, nxt_http_request_t *r,
+    nxt_router_access_log_ctx_t *ctx, nxt_router_access_log_format_t *format)
+{
+    u_char                          *p;
+    size_t                          size;
+    nxt_int_t                       ret;
+    nxt_str_t                       str;
+    nxt_uint_t                      i;
+    nxt_conf_value_t                *value;
+    nxt_router_conf_t               *rtcf;
+    nxt_router_access_log_member_t  *member;
+
+    rtcf = r->conf->socket_conf->router_conf;
+
+    value = nxt_conf_create_object(r->mem_pool, format->nmembers);
+    if (nxt_slow_path(value == NULL)) {
+        return NXT_ERROR;
+    }
+
+    for (i = 0; i < format->nmembers; i++) {
+        member = &format->member[i];
+
+        if (nxt_tstr_is_const(member->tstr)) {
+            nxt_tstr_str(member->tstr, &str);
+
+        } else {
```

Comment:
Looks like `ret` could be declared here...

Reply:
`member` could be declared in the for loop...

Reply:
Good to know, thanks for the suggestions.
I'm fine with both styles. And I'm used to declaring variables at the beginning of functions, so I'd like to keep this way.

Reply:
Well, it is more than about style, it's good programming [practice](https://wiki.sei.cmu.edu/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions), which is why I won't stop suggesting it ;)

Note: declaring variables at start of scope is _not_ the same as declaring variables randomly throughout the function, which I certainly _wouldn't_ suggest to do!

Reply:
I think both ways have pros and cons. Declaring at the start makes all variables visible at a glance and makes reuse easier.

Take this function as an example:
```
static nxt_int_t
nxt_router_access_log_json(nxt_task_t *task, nxt_http_request_t *r,
    nxt_router_access_log_ctx_t *ctx, nxt_router_access_log_format_t *format)
{
    nxt_uint_t                      i;
    nxt_conf_value_t                *value;
    nxt_router_access_log_member_t  *member;

    value = nxt_conf_create_object(r->mem_pool, format->nmembers);
    if (nxt_slow_path(value == NULL)) {
        return NXT_ERROR;
    }

    for (i = 0; i < format->nmembers; i++) {
        member = &format->member[i];

        ...

        nxt_conf_set_member_string(value, &member->name, &str, i);
    }

    ...

    return NXT_OK;
}
```
I feel it's common to write like it. If we asked someone to totally follow something like the above "practice", it looks like that:
```
static nxt_int_t
nxt_router_access_log_json(nxt_task_t *task, nxt_http_request_t *r,
    nxt_router_access_log_ctx_t *ctx, nxt_router_access_log_format_t *format)
{
    nxt_conf_value_t  *value;

    value = nxt_conf_create_object(r->mem_pool, format->nmembers);
    if (nxt_slow_path(value == NULL)) {
        return NXT_ERROR;
    }

    for (nxt_uint_t i = 0; i < format->nmembers; i++) {
        nxt_router_access_log_member_t  *member;

        member = &format->member[i];

        ...

        nxt_conf_set_member_string(value, &member->name, &str, i);
    }

    ...

    return NXT_OK;
}
```
From the point of code style, both ways look fine.
It doesn't seem to make a difference in terms of whether it's more buggy or not.

Reply:
For me the second version is more concise.

It also has the advantage that suppose you _mistakenly_ tried to do something with `member` after the for loop, the compiler will shout at you... unlike the first version where all kinds of mayhem may ensue...

So I will continue pushing it (even if I do sound like a broken record) as in the current climate anything that helps write _safer_ code is a _good_ thing...

Reply:
Thanks for caring about code safety. I'm comfortable with my current style of declaring at the beginning. Both ways can be safe when written carefully, especially with tests, so I'd like to keep my approach.

Reply:
> Thanks for caring about code safety. I'm comfortable with my current style of declaring at the beginning. Both ways can be safe when written carefully

Careful you don't fall into the fallacy that you won't make mistakes no matter how careful you are.

I don't care _who_ you are, if you say you _never_ write buggy code, then you are either outright lying or delusional!

> , especially with tests, so I'd like to keep my approach.

Tests are good but not infallible...

Your choice, I'll just keep bugging you ;) (though I won't bring it up again in this PR)

Reply:
> I don't care who you are, if you say you never write buggy code, then you are either outright lying or delusional!

I discussed it based on the specific code which I listed above. Please do not expand the scope and apply one "principle" to all situations.
I didn't say I never write buggy code. I mean the way I used in `nxt_router_access_log_json()` didn't introduce issues, let me know if it's possible.

Furthur example:
```
int array_sum(int a[], int n)
{
    int i, sum;

    sum = 0;

    for (i = 0; i < n; i++) {
        sum += a[i];
    }

    return sum;
}
```
I don't think it should be like that, but I'm fine if anyone likes the way.
```
int array_sum(int a[], int n)
{
   int sum;

    sum = 0;

    for (int i = 0; i < n; i++) {
        sum += a[i];
    }

    return sum;
}
```

Reply:
> I discussed it based on the specific code which I listed above. Please do not expand the scope and apply one "principle" to all situations.

I _am_ talking generally (not specific examples)...

Reducing variable scope is just good programming practice... ipso facto...

Reply:
Take `nxt_router_access_log_json()` as an example, `i` and `member` are only used in the `for` sentence, I can't see benefits to force their declaration within it.
I would be confused if it can introduce issues in this way, I'm only talking about the function since we are reviewing it.

Reply:
> Take `nxt_router_access_log_json()` as an example, `i` and `member` are only used in the `for` sentence,

Yes, that's exactly the point! They are _only_ used in the for loop so that is where they should be _scoped_.

> I can't see benefits to force their declaration within it.

I already gave an example of a benefit...

> I would be confused if it can introduce issues in this way, I'm only talking about the function since we are reviewing it.

Never underestimate the propensity for introducing bugs.

Reply:
I wonder if a function is as small as array_sum(), should it follow the practice?

Reply:
Certainly no harm... and it's more concise to read...